sqlglot/optimizer/expand_join_constructs.py

```diff
@@ -0,0 +1,25 @@
+from sqlglot import exp
+
+
+def expand_join_constructs(expression: exp.Expression) -> exp.Expression:
+    """
+    Replace "join constructs" (*) by equivalent SELECT * subqueries.
+
+    Example:
+    >>> import sqlglot
+    >>> expression = sqlglot.parse_one("SELECT * FROM (tbl1 AS tbl1 JOIN tbl2 AS tbl2 ON id1 = id2) AS tbl")
+    >>> expand_join_constructs(expression).sql()
+    'SELECT * FROM (SELECT * FROM tbl1 AS tbl1 JOIN tbl2 AS tbl2 ON id1 = id2) AS tbl'
+
+    (*) See section 7.2.1.2 in https://www.postgresql.org/docs/current/queries-table-expressions.html
+    """
+
+    def _expand_join_constructs(expression: exp.Expression) -> exp.Expression:
+        if isinstance(expression, exp.Subquery):
+            unnested = expression.unnest()
+            if isinstance(unnested, exp.Table):
+                expression.this.replace(exp.select("*").from_(unnested.copy(), copy=False))
```

Comment:
why do you need to do unnest.copy() ?

Reply:
We need to copy because `expression.this` is the same as `unnested` in the following case:
```
SELECT * FROM (tbl AS tbl)
```
If we don't copy, then `from_` will set `unnested`'s parent to the `exp.From` expression and then `replace` will mess it up again. The order of operations is not well defined (I tested this too, the transformation fails).